### PR TITLE
remove fallback to default domain

### PR DESCRIPTION
**Pull request by apanicker-nflx**
Removed the fallback to the default domain if an active domain is not found.
Changed this such that it falls back to the lowest priority domain specified in the list.

### CODE REVIEW

File: core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java [hunk 1]
```diff
@@ -1062,16 +1062,18 @@ private void setTaskDomains(List<Task> tasks, Workflow wf) {
...
+        // In sequence check if any worker has polled for last 30 seconds, if so that is the Active domain
+
         return Arrays.stream(domains)
                 .map(domain -> executionDAO.getPollData(taskType, domain.trim()))
                 .filter(Objects::nonNull)
                 .filter(validateLastPolledTime)
                 .findFirst()
                 .map(PollData::getDomain)
-                .orElse(null);
+                .orElse(domains[domains.length - 1].trim());
```
Comment (kishorebanala) [inline core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java]: Would it be possible to throw ArrayIndexOutOfBoundsException when domains length is 0?

File: common/src/main/java/com/netflix/conductor/common/metadata/tasks/PollData.java [hunk 2]
```diff
@@ -102,8 +107,7 @@ public synchronized boolean equals(Object obj) {
 		} else if (!queueName.equals(other.queueName))
 			return false;
 		if (workerId == null) {
-			if (other.workerId != null)
-				return false;
+			return other.workerId == null;
```
Comment (alexnetflix) [inline common/src/main/java/com/netflix/conductor/common/metadata/tasks/PollData.java]: consideration: I like Jave 7's  "if(!Objects.equals(x,y) return false" statements, a bit more concise, handles nulls, and leaves less room for error if somebody adds more checks at the end of the method for example and other.workerId == null returns true

Comment (cyzhao) [reply common/src/main/java/com/netflix/conductor/common/metadata/tasks/PollData.java]: I agree, let's regenerate equals() and hashCode(), also remove synchronized keyword from these methods.

File: core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java [hunk 3]
```diff
@@ -1062,16 +1062,18 @@ private void setTaskDomains(List<Task> tasks, Workflow wf) {
         }
     }
 
-    private String getActiveDomain(String taskType, String[] domains) {
+    @VisibleForTesting
+    String getActiveDomain(String taskType, String[] domains) {
```
[the remark this answers is not in the document]
Comment (apanicker-nflx) [reply core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java]: https://github.com/Netflix/conductor/pull/864/files#diff-a28de3002a75934bb7c693d03b674144L1036
This is the only method accessing this function, null values cannot be passed as domains here. I will update the javadoc with the same.

File: core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java (hunk 1 above)
Comment (apanicker-nflx) [reply core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java]: As mentioned above, it is guaranteed that the domains array is atleast of size 1 when passed as param into the function. I will update the javadoc.

File: core/src/test/java/com/netflix/conductor/core/execution/TestWorkflowExecutor.java [hunk 4]
```diff
@@ -522,7 +526,27 @@ public void testRetryWorkflow() {
         assertEquals(2, updateWorkflowCalledCounter.get());
         assertEquals(7, updateTasksCalledCounter.get());
         assertEquals(1, removeQueueEntryCalledCounter.get());
+    }
+
+    @Test
+    public void testGetActiveDomain() {
```
[the remark this answers is not in the document]
Comment (apanicker-nflx) [reply core/src/test/java/com/netflix/conductor/core/execution/TestWorkflowExecutor.java]: Added single empty domain test case.

File: core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (apanicker-nflx) [reply core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java]: Added checks and test cases

File: core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java (hunk 1 above)
Comment (apanicker-nflx) [reply core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java]: Added checks and tests

**Pull request merged**